### PR TITLE
Feature: linear cross entropy fusion

### PR DESCRIPTION
**Pull request by Jianbing-D**
# What does this PR do ?

This PR introduces an implementation that fuses Linear Layer of `lm_head` and Cross-Entropy, in order to avoid materializing the intermediate `logits` tensor, helping reducing memory footprint.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share discuss a design-doc with the team.  

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [x] I have added relevant unit tests
- [x] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Comment by Jianbing-D** [comment]
# Details about this feature
Training LLM typically involves a two-stage pipeline at the output layer: hidden states are projected into vocabulary `logits` via a linear transformation (lm_head Layer), followed by Cross-Entropy loss computation against target tokens. While conceptually simple, such workflow incurs substantial overhead. The intermediate logits tensor, with dimension proportional to `batch size, sequence length, and vocabulary size`, must be fully materialized in GPU memory, even though only one target token per position is ultimately used. This leads to significant memory footprint and bandwidth consumption, limiting scalability and slowing training throughput. The following code snippet might better illustrate that workflow:
```python
hidden_state = xxx # shape = [batch, seqlen, dim]
weight = lm_head.weight # shape = [vocabsize, dim]
labels = xxx # shape = [batch, seqlen]

logits = hidden_state @ weight.T
loss = torch.nn.functional.cross_entropy(logits, labels, reduction="none")
```

On top of the `local logit` tensor, other techniques might need some other intermediate buffers for collecting full information across all GPUs. For example, the following snippet is a TP compatible layer, comprised of torch native ops:
```python
tp_rank = xxx
tp_world_size = xxx

logits = hidden @ weight.T

whole_logits = torch.empty(
    (logits.shape[0], logits.shape[-1] * tp_world_size),
     dtype=logits.dtype,
     device=logits.device,
)
whole_logits_ref = [
     whole_logits[..., i * logits.shape[-1] : (i + 1) * logits.shape[-1]]
     for i in range(tp_world_size)
]
dist.all_gather(whole_logits_ref, logits, group=tp_group)

logprobs = torch.nn.functional.cross_entropy(
    whole_logits.view(-1, whole_logits.shape[-1]), labels.view(-1), reduction=reduction
)
```

By fusing `Linear` and `Cross-Entropy` into one single operation, this PR could help avoid materializing the intermediate `logit` tensor.
```python
hidden_state = xxx # shape = [batch, seqlen, dim]
weight = lm_head.weight # shape = [vocabsize, dim]
labels = xxx # shape = [batch, seqlen]

loss = linear_cross_entropy(hidden_state, weight, labels, reduction="none")
```

which could help reduce `2bsv` memory footprints **AT LEAST**.
+ in the forward pass, no need to materializing `logit` tensor, whose shape is `[batch, seqlen, vocabsize]`
+ in the backward pass, no need to materializing `grad of logit` tensor, whose shape is also `[batch, seqlen, vocabsize]`

## functionalities ##
```python
def linear_cross_entropy(
    hidden: torch.Tensor,
    weight: torch.Tensor,
    labels: torch.Tensor,
    tp_group: typing.Optional[torch.distributed.ProcessGroup] = None,
    reduction: str = "mean",
    ignore_index: int = -100,
    sequence_parallel: bool = False,
) -> torch.Tensor
```
+ The input tensor is `BF16 or FP16` format, and will conduct accumulation and other logics in `FP32` format, avoiding precision problem.
+ It supports `Data-Parallel`, `Tensor-Parallel along vocabsize`, and `Sequence-Parallel along seqlen`.
  1. when `tp_group is None` it works in DP mode
  2. when `tp_group is not None, and sequence_parallel is False`, it works in TP mode
  3. when `tp_group is not None, and sequence_parallel is True`, it works in SP mode
+ It supports specifying `ignore_idex` as `what native torch cross-entropy does`.
+ It supports specifying `reduction` method as `what native torch cross-entropy does`.
+ It is optimized for latest NVIDIA Blackwell GPUs.

## Performance and Storage ##
In DP mode, this PR could lead to perf boost and storage reduction in the following config:
image

You may try the following steps to reproduce it:
```shell
# start a Megatron image on GB200
$ pip install nvidia-cutlass-dsl==4.2.1
$ pip install PyGithub
$ pytest -s -v tests/unit_tests/fusions/test_fused_linear_cross_entropy.py
$ torchrun --nproc_per_node=4 --nnodes=1 -m pytest -s -v tests/unit_tests/fusions/test_fused_linear_cross_entropy.py
```

**Comment by Jianbing-D** [comment]
We have already done some verification experiments on real LLM models to verify the accuracy / convergency and storage reduction while utilizing this feature. @shjwudp Please comment here.

**Comment by kvareddy** [comment]
Did you try out storing the logits in bf16?  It could save lot of memory. Not sure if we need this fusion.

**Comment by shjwudp** [comment]
> We have already done some verification experiments on real LLM models to verify the accuracy / convergency and storage reduction while utilizing this feature. @shjwudp Please comment here.

## Convergence Test

We conducted tests on **Llama3-8B** and **DSv3-Proxy** (with the MTP layer). The convergence performance of the **Linear Cross-Entropy Fusion** method was comparable to the **M-Core baseline**.

screenshot-20251111-190423

## Memory Saving Test

Compared with **TE Cross-Entropy Fusion**, the proposed **Linear Cross-Entropy Fusion (this MR)** achieved a memory reduction of approximately **3.9 GB**.

**Cross-Entropy Fusion**:
image

**Linear-Cross-Entropy Fusion (This MR)**:
image

**Comment by shjwudp** [comment]
> Did you try out storing the logits in bf16? It could save lot of memory. Not sure if we need this fusion.

@kvareddy Yes, we tested bf16 activations (current TE cross-entropy fusion), and when the vocabulary size is large, the Linear-Cross-Entropy Fusion remains more memory-efficient compared to the CE fusion. Moreover, the linear-CE fusion preserves the precision of the logits better, as it avoids down-cast and up-cast during computation.

**Comment by lhb8125** [comment]
> Did you try out storing the logits in bf16? It could save lot of memory. Not sure if we need this fusion.

@kvareddy Jack has given a good case from the viewpoint of memory saving and convergence. Theoretically, the logits in bf16 still could become a bottleneck in the fine-grained moe models, which have a high memory pressure. 

With TP&SP&selective recomputation, the memory footprint of a layer is `bsh*(34)/tp`, while for the logit, the memory footprint is `bsv/tp`, which is roughly equivalent to half of the layer. For deepseek-v3, the best pp layout as of now is `Et|(tt|)*30mL`, the MTP layer contains a full model with 1 transformer layer, so the last pp stage could be the bottleneck of memory footprint, considering the large vocab size and optimizer states. Moreover, the memory footprint of a layer could be further reduced by fine-grained recomputation and offloading.

cc @yanring

**Comment by Skylion007** [comment]
Is this the dtype aware cut cross entropy kernel that helps with large vocabularies on already pretrained models or just the less specialized liger variant?

File: megatron/core/fusions/linear_cross_entropy/blackwell/entry.py [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+from megatron.core.fusions.linear_cross_entropy.blackwell import triton as triton_kernels
+
+
+class FwdConfig:
+    """
+    The configuration for the forward pass.
+    """
+
+    _dedicated_stream: torch.cuda.Stream = None
+    _dedicated_events: typing.List[torch.cuda.Event] = list()
```
Comment (Skylion007) [inline megatron/core/fusions/linear_cross_entropy/blackwell/entry.py]: Thee configs will have singleton collectives which is usually not desired. Better to have them be data classes and use dataclass.field factor kwargs to ensure every instance has its own unique member collection instead of all instances pointing to the same list. Current behavior is bugprone

File: megatron/core/fusions/linear_cross_entropy/utils.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    """
+    Enum for the reduction method of cross entropy.
+    """
+
+    kNone = 0
+    kSum = 1
+    kMean = 2
+
+
+def str_to_reduction_enum(reduction: str) -> int:
```
Comment (Skylion007) [inline megatron/core/fusions/linear_cross_entropy/utils.py]: Input can be a literal for static typing

File: megatron/core/fusions/linear_cross_entropy/utils.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
+# Copyright (c) 2025, NVIDIA CORPORATION.  All rights reserved.
+
+from dataclasses import dataclass
+
+
+@dataclass
```
Comment (Skylion007) [inline megatron/core/fusions/linear_cross_entropy/utils.py]: Why have this be a dataclass instead of an enum?

File: megatron/core/fusions/linear_cross_entropy/utils.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    elif reduction == "sum":
+        _enum = EntropyReductionEnum.kSum
+    elif reduction == "mean":
+        _enum = EntropyReductionEnum.kMean
+    else:
+        raise ValueError(f"Invalid reduction: {reduction}")
+    return _enum
+
+
+@dataclass
```
Comment (Skylion007) [inline megatron/core/fusions/linear_cross_entropy/utils.py]: Same here, just subclass enum instead.

File: megatron/training/arguments.py [hunk 5]
```diff
@@ -2176,6 +2176,10 @@ def _add_training_args(parser):
                        dest='bias_swiglu_fusion')
     group.add_argument('--use-fused-weighted-squared-relu', action='store_true',
                        help='Use fused weighted squared relu when using MoE.')
+    group.add_argument('--linear-cross-entropy-fusion', action='store_true',
```
Comment (Skylion007) [inline megatron/training/arguments.py]: We already have cross entropy fusion and cross entropy impl args. Why don't we just role into an option there instead of adding yet another new arg (which could conflict with the other two). Rapidly falling into the Boolean trap.

**Comment by Skylion007** [comment]
Linking this issue: https://github.com/NVIDIA/Megatron-LM/issues/1738

File: megatron/training/arguments.py (hunk 5 above)
Comment (shjwudp) [reply megatron/training/arguments.py]: Good point! Will merge those two flags.

File: megatron/core/fusions/linear_cross_entropy/blackwell/entry.py (hunk 1 above)
Comment (Jianbing-D) [reply megatron/core/fusions/linear_cross_entropy/blackwell/entry.py]: Thanks for your suggestion. Have updated those definitions.

File: megatron/core/fusions/linear_cross_entropy/utils.py (hunk 2 above)
Comment (Jianbing-D) [reply megatron/core/fusions/linear_cross_entropy/utils.py]: Done.

File: megatron/core/fusions/linear_cross_entropy/utils.py (hunk 3 above)
Comment (Jianbing-D) [reply megatron/core/fusions/linear_cross_entropy/utils.py]: Done, updated to be a subclass of Enum.

File: megatron/core/fusions/linear_cross_entropy/utils.py (hunk 4 above)
Comment (Jianbing-D) [reply megatron/core/fusions/linear_cross_entropy/utils.py]: Done.

**Comment by kvareddy** [comment]
Can we make this feature optional? Also, can we move the kernels to TE and use them in the Megatron core?

File: megatron/core/models/common/language_module/language_module.py [hunk 6]
```diff
@@ -117,6 +118,68 @@ def check_and_set_env_variable(
             check_and_set_env_variable("NVTE_FUSED_ATTN", 1, AttnBackend.auto)
             check_and_set_env_variable("NVTE_UNFUSED_ATTN", 1, AttnBackend.auto)
 
+    def compute_language_model_loss_without_logits(
```
Comment (jaredcasper) [inline megatron/core/models/common/language_module/language_module.py]: This function name doesn't really make sense. First, if the fusion argument isn't true, then logits are created. Second, logits are still used they just aren't materialized (for what I can gather). You can't compute cross entropy loss without logits. Should this just fully replace `compute_language_model_loss`?

Comment (shjwudp) [reply megatron/core/models/common/language_module/language_module.py]: I agree — the name of this function doesn’t seem very appropriate. Would it be better to call it `compute_output_layer_and_language_model_loss` instead?

As for `compute_language_model_loss`, I think it should remain a separate function. In some cases, we need to compute the logits first and use them to compute the loss.

**Comment by shjwudp** [comment]
@sanandaraj5597 Would you be willing to review this MR? Since you previous contributions to cross-entropy fusion, we would love to hear your feedback, thank you!

**Comment by shjwudp** [comment]
> Can we make this feature optional? Also, can we move the kernels to TE and use them in the Megatron core?

Hi @kvareddy, this feature is currently optional and can be controlled via `--cross-entropy-fusion-impl linear`; it is not enable by default. The linear cross-entropy implementation is built using Triton and the CUTE DSL. Given that some of Megatron’s fused operators are also implemented in Triton, this approach isn’t really an exception. If the linear cross-entropy proves to be sufficiently general, I think it could be moved into TE in the future.

File: megatron/core/transformer/transformer_config.py [hunk 7]
```diff
@@ -292,6 +292,9 @@ class TransformerConfig(ModelParallelConfig):
     fused_single_qkv_rope: bool = False
     """If set, avoid splitting QKV before ROPE forward and avoid concatenating ROPE dgrads."""
 
+    linear_cross_entropy_fusion: bool = False
```
Comment (yaox12) [inline megatron/core/transformer/transformer_config.py]: Not needed any more.

Comment (shjwudp) [reply megatron/core/transformer/transformer_config.py]: Thank you for pointing it out, I will delete it.

File: megatron/core/models/common/language_module/language_module.py (hunk 6 above)
Comment (Jianbing-D) [reply megatron/core/models/common/language_module/language_module.py]: Marked as resolved, as @shjwudp has already changed it to `compute_output_layer_and_language_model_loss`

File: megatron/core/transformer/transformer_config.py (hunk 7 above)
Comment (Jianbing-D) [reply megatron/core/transformer/transformer_config.py]: Marked as resolved, as @shjwudp has already delete it.

**Comment by Skylion007** [comment]
Has this been tested with large vocab sizes? We observed numerics issue with our port of a liger kernel into our codebase a few months ago when the vocab size of the model exceeded 250k (needed to upcast some accumulators if I recall). Actually worked fine at Llama vocab sizes, may want to double check: @shjwudp . Led to a subtle convergence issue IIRC.

**Comment by shjwudp** [comment]
> Has this been tested with large vocab sizes? We observed numerics issue with our port of a liger kernel into our codebase a few months ago when the vocab size of the model exceeded 250k (needed to upcast some accumulators if I recall). Actually worked fine at Llama vocab sizes, may want to double check: @shjwudp . Led to a subtle convergence issue IIRC.

I indeed have not conducted experiments on the 256K vocabulary. As I understand it, the Linear-CE fusion algorithm is fully consistent with the standard Cross-Entropy formulation, so there shouldn’t be any numerical issues even with a larger vocabulary size. You could run a test with the 256K vocab to confirm — thanks for letting us know.

cc @Jianbing-D @kunlunl  Please correct me if I’m mistaken.

**Comment by Skylion007** [comment]
> Is this the dtype aware cut cross entropy kernel that helps with large vocabularies on already pretrained models or just the less specialized liger variant?

Just wanted to double check this as this could massively reduce memory for finetuning (and improve speed in that case for LLMs) while being bit wise numerically equivalent: https://github.com/apple/ml-cross-entropy

**Comment by Jianbing-D** [comment]
> > Is this the dtype aware cut cross entropy kernel that helps with large vocabularies on already pretrained models or just the less specialized liger variant?
> 
> Just wanted to double check this as this could massively reduce memory for finetuning (and improve speed in that case for LLMs) while being bit wise numerically equivalent: https://github.com/apple/ml-cross-entropy

Hi, this kernel follows this scheme to do its calculation:
```python
logits = hidden_state.to(torch.float32) @ weight.T.to(torch.float32)
loss = torch.nn.functional.cross_entropy(logits, labels, reduction="none")
```
The `logits` tensor will be eliminated since the `GEMM` process has been fused into the `CrossEntropy`. As you might've noticed, the `GEMM` process will produce `Float32`, and the subsequent `CrossEntropy` will also be conducted in `Float32` format, to assure the numerical precision. 

We can analyze this operation in two separate parts, `forward pass` and `backward pass`. And you can refer to previous experiments result to check the latency speedup and memory shrinkage. For implementation detail:
+ `forward pass` will totally eliminate the `logits` tensor, which leads to significant memory reduction;
+ `backward pass` also don't need a full `dlogits` tensor, which leads to significant memory reduction as well.
+ both `logits` and `dlogits` tensor scale with `BatchSize`, `SequenceLength`, and `VocabSize`. The more memory savings, the larger those parameters increase.
